crates/pipeline_manager/src/db/mod.rs

```diff
@@ -717,8 +989,154 @@ impl Storage for ProjectDB {
         }
     }
 
+    /// Version the current pipeline object and all state reachable from it.
+    ///
+    /// We store the last revision number in the pipeline object itself.
+    async fn create_pipeline_revision(
+        &self,
+        tenant_id: TenantId,
+        pipeline_id: PipelineId,
+    ) -> AnyResult<Revision> {
+        let mut client = self.pool.get().await?;
+        let txn = client.transaction().await?;
+
+        // TODO: Ideally these queries happen in the same transaction `txn`.
```

Comment:
It seems like you are using a transaction though?

Reply:
Yep, the comment is for `pipeline_is_committable` which should run in the same transaction... but then you need to pass down the transaction to all those calls (or duplicate all those queries made there which seems bad)... Instead of duplicating all the code I can do a pass over the API so we can pass in a transaction... we also duplicate the queries for program, pipeline etc. a few times (for name, uuid retreival etc), and we're not using prepared statements. So was hoping to fix that then too.